dev/tools/gen_keycodes/lib/physical_key_data.dart

```diff
@@ -0,0 +1,366 @@
+    final Map<String, List<int>> nameToGlfwKeyCodes = _readGlfwKeyCodes(glfwHeaderFile, glfwNameMap);
+    final Map<String, PhysicalKeyEntry> data = _readHidEntries(
+      chromiumHidCodes,
+      nameToAndroidScanCodes,
+      nameToGlfwKeyCodes,
+    );
+    final List<MapEntry<String, PhysicalKeyEntry>> sortedEntries = data.entries.toList()..sort(
+      (MapEntry<String, PhysicalKeyEntry> a, MapEntry<String, PhysicalKeyEntry> b) =>
+        PhysicalKeyEntry.compareByUsbHidCode(a.value, b.value),
+    );
+    data
+      ..clear()
+      ..addEntries(sortedEntries);
+    return PhysicalKeyData._(data);
+  }
+
+  /// Parses the given JSON data and populates the data structure from it.
+  factory PhysicalKeyData.fromJson(Map<String, dynamic> contentMap) {
+    final Map<String, PhysicalKeyEntry> data = <String, PhysicalKeyEntry>{};
+    for (final MapEntry<String, dynamic> jsonEntry in contentMap.entries) {
+      final PhysicalKeyEntry entry = PhysicalKeyEntry.fromJsonMapEntry(jsonEntry.value as Map<String, dynamic>);
+      data[entry.name] = entry;
+    }
+    return PhysicalKeyData._(data);
+  }
+
+  PhysicalKeyData._(this._data);
+
+  /// Find an entry from name, or null if not found.
+  PhysicalKeyEntry? tryEntryByName(String name) {
+    return _data[name];
+  }
+
+  /// Find an entry from name.
+  ///
+  /// Asserts if the name is not found.
+  PhysicalKeyEntry entryByName(String name) {
+    final PhysicalKeyEntry? entry = tryEntryByName(name);
+    assert(entry != null,
+        'Unable to find logical entry by name $name.');
+    return entry!;
+  }
+
+  /// All entries.
+  Iterable<PhysicalKeyEntry> get entries => _data.values;
+
+  // Keys mapped from their names.
+  final Map<String, PhysicalKeyEntry> _data;
+
+  /// Converts the data structure into a JSON structure that can be parsed by
+  /// [PhysicalKeyData.fromJson].
+  Map<String, dynamic> toJson() {
+    final Map<String, dynamic> outputMap = <String, dynamic>{};
+    for (final PhysicalKeyEntry entry in _data.values) {
+      outputMap[entry.name] = entry.toJson();
+    }
+    return outputMap;
+  }
+
+  /// Parses entries from Androids Generic.kl scan code data file.
+  ///
+  /// Lines in this file look like this (without the ///):
+  /// key 100   ALT_RIGHT
+  /// # key 101 "KEY_LINEFEED"
+  ///
+  /// We parse the commented out lines as well as the non-commented lines, so so
+  /// that we can get names for all of the available scan codes, not just ones
+  /// defined for the generic profile.
+  ///
+  /// Also, note that some keys (notably MEDIA_EJECT) can be mapped to more than
+  /// one scan code, so the mapping can't just be 1:1, it has to be 1:many.
+  static Map<String, List<int>> _readAndroidScanCodes(String keyboardLayout, String nameMap) {
+    final RegExp keyEntry = RegExp(r'#?\s*key\s+([0-9]+)\s*"?(?:KEY_)?([0-9A-Z_]+|\(undefined\))"?\s*(FUNCTION)?');
```

Comment:
I would suggest using named groups when you have multiple groups in a RegExp, so that it is self documenting, and easier to maintain, since you don't have as much of a chance to accidentally change the group index by changing the regular expression.  Something like this:
```suggestion
    final RegExp keyEntry = RegExp(r'#?\s*key\s+(?<id>[0-9]+)\s*"?(?:KEY_)?(?<name>[0-9A-Z_]+|\(undefined\))"?\s*(?<function>FUNCTION)?');
```

And then the `Match` should be changed to the newer `RegExpMatch` type, and you can access the groups with `match.namedGroup('name')`, etc.

I'd recommend doing that for all the non-trivial `RegExp`s here.

Reply:
I changed the 2 REs in `physical_key_data` and 1 RE in `logical_key_data`.